Add `transform.hsl()`

itzpr3d4t0r opened the pull request:
Closes #2310.
This PR implements a general purpose function that lets users modify Surface's hue, saturation and lighness values.
Still a draft as it's missing docs, tests and possibly improved performance / fixes.

PurityLake commented:
Looks fairly good, only thing really needed now is tests, and it should be good to go.

itzpr3d4t0r commented:
Bulk of the work is done. I changed the algorithm for a 19% perf improvement, added tests and docs. 
What is still missing is checking that the algorithm always works without strange quirks, possibly adding more tests (for the various bit depths) and fixing the reason why the tests are failing, which is a one-value-off issue that's a combination of factors.

yunline commented on src_c/transform.c:
```diff
@@ -2118,6 +2118,225 @@ surf_grayscale(PyObject *self, PyObject *args, PyObject *kwargs)
+    }
+    h /= 360.0;
+
+    newsurf = modify_hsl(surfobj, surfobj2, h, s, l);
```
Shall we release the GIL here?

MyreMylar commented on src_c/transform.c:
```diff
@@ -2118,6 +2118,225 @@ surf_grayscale(PyObject *self, PyObject *args, PyObject *kwargs)
+    }
+    h /= 360.0;
+
+    newsurf = modify_hsl(surfobj, surfobj2, h, s, l);
```
```suggestion
    pgSurface_Lock(surfobj);
    Py_BEGIN_ALLOW_THREADS;
    newsurf = modify_hsl(surfobj, surfobj2, h, s, l);
    Py_END_ALLOW_THREADS;
    pgSurface_Unlock(surfobj);
```

Do you mean something like this?

MyreMylar replied:
This was meant in reply to @yunline

yunline replied:
> Do you mean something like this?

Yes that is what I mean.

itzpr3d4t0r replied:
I added the thread stuff but seems like it led to errors. I don't know wy tbh.

MyreMylar commented on src_c/transform.c:
```diff
@@ -2118,6 +2118,225 @@ surf_grayscale(PyObject *self, PyObject *args, PyObject *kwargs)
+        }
+    }
+
+    SDL_UnlockSurface(newsurf);
```
Don't we need to lock this earlier to unlock it here?

itzpr3d4t0r replied:
I followed what i saw in grayscale, not sure maybe since we create a new surface it's already locked? gotta take a better look since i don't really remember...

MyreMylar reviewed (changes requested):
Overall this works well, I just have a couple of questions about surface locking and unlocking and as this seems to be done in several different ways all over the transform module. Looking up the python C API it seems to say:

> Note Calling system I/O functions is the most common use case for releasing the GIL, but it can also be useful before calling long-running computations which don’t need access to Python objects, such as compression or cryptographic functions operating over memory buffers. For example, the standard [zlib](https://docs.python.org/3/library/zlib.html#module-zlib) and [hashlib](https://docs.python.org/3/library/hashlib.html#module-hashlib) modules release the GIL when compressing or hashing data.

Which I think all our transforms would qualify for.

yunline commented on src_c/transform.c:
```diff
@@ -2118,6 +2118,227 @@ surf_grayscale(PyObject *self, PyObject *args, PyObject *kwargs)
+    }
+
+    if (newsurf->w != surf->w || newsurf->h != surf->h) {
+        return (SDL_Surface *)(RAISE(
```
You can't `RAISE` here, since `PyErr_SetString()` needs GIL.
This may be the reason why the tests failed

yunline replied:
Try move the `RAISE`s out of this function

MyreMylar commented on src_c/transform.c:
```diff
@@ -2211,6 +2211,216 @@ surf_grayscale(PyObject *self, PyObject *args, PyObject *kwargs)
+static void
+modify_hsl(SDL_Surface *surf, SDL_Surface *dst, float h, float s, float l)
+{
+    int x, y;
```
```suggestion
    int surf_locked = 0;
    if (SDL_MUSTLOCK(surf)) {
        if (SDL_LockSurface(surf) == 0) {
            surf_locked = 1;
        }
    }
    int x, y;
```

MyreMylar commented on src_c/transform.c:
```diff
@@ -2211,6 +2211,216 @@ surf_grayscale(PyObject *self, PyObject *args, PyObject *kwargs)
+        }
+    }
+
+    SDL_UnlockSurface(dst);
```
```suggestion
    if (surf_locked) {
        SDL_UnlockSurface(dst);
    }
```

MyreMylar reviewed (commented):
I spent a bit of time looking into surface locking in SDL. Basically it is only needed when the surface has been RLE encoded.

As far as I can tell (though [there is an open bug](https://github.com/libsdl-org/SDL/issues/5594) about it, but I agree with the last comment), `MUST_LOCK(surf)` tells us whether the surface needs locking - which, if true, is something we should do before any direct pixel access.

I've added two suggestions to the code here that will lock the surface before accessing the pixels and then gate the call to unlock only if the surface is already locked.

Untested, but you could probably create a surface with RLE - blitting it to activate the RLE encoding on the surface - and then try running the hsl function post blit to create a good test of locking/unlocking.

MyreMylar commented on src_c/transform.c:
```diff
@@ -2211,6 +2211,224 @@ surf_grayscale(PyObject *self, PyObject *args, PyObject *kwargs)
+static void
+modify_hsl(SDL_Surface *surf, SDL_Surface *dst, float h, float s, float l)
+{
+    int surf_locked = 0;
+    if (SDL_MUSTLOCK(surf)) {
+        if (SDL_LockSurface(surf) == 0) {
+            surf_locked = 1;
+        }
+    }
```
```suggestion
    int surf_locked = 0;
    if (SDL_MUSTLOCK(surf)) {
        if (SDL_LockSurface(surf) == 0) {
            surf_locked = 1;
        }
    }
    int dest_locked = 0;
    if (SDL_MUSTLOCK(dst)) {
        if (SDL_LockSurface(dst) == 0) {
            dst_locked = 1;
        }
    }
```

yunline reviewed (approved):
LGTM👍

Starbuck5 commented on src_c/transform.c:
```diff
@@ -2211,6 +2211,233 @@ surf_grayscale(PyObject *self, PyObject *args, PyObject *kwargs)
+    SDL_Surface *dst, *src;
+    float h = 0, s = 0, l = 0;
+
+    static char *keywords[] = {"surface",   "hue",          "saturation",
```
The keywords don't match the docs.

itzpr3d4t0r replied:
Thanks!

Starbuck5 commented on src_c/transform.c:
```diff
@@ -2211,6 +2211,233 @@ surf_grayscale(PyObject *self, PyObject *args, PyObject *kwargs)
+        return NULL;
+    }
+
+    if (h > 0) {
+        while (h > 360) {
+            h -= 360;
+        }
+    }
+    else if (h < 0) {
+        while (h < -360) {
+            h += 360;
+        }
+    }
```
Isn't this equivalent to `h %= 360` ?

itzpr3d4t0r replied:
Thanks!

Starbuck5 commented on src_c/transform.c:
```diff
@@ -2211,6 +2211,233 @@ surf_grayscale(PyObject *self, PyObject *args, PyObject *kwargs)
+            PyExc_ValueError,
+            "Destination surface must be the same size as source surface.");
+    }
+    if (src->format->BytesPerPixel != dst->format->BytesPerPixel) {
+        return RAISE(PyExc_ValueError,
+                     "Source and destination surfaces need the same format.");
+    }
```
If you want to check format, check src->format->format. Them being the same bytes per pixel does not mean they are compatible at all (yes I know other methods in transform do this, but best not to replicate).

itzpr3d4t0r replied:
Thanks!

itzpr3d4t0r commented:
I have an update to share. I’ve managed to achieve a substantial performance improvement of approximately 43% for this Pull Request when dealing with 4bpp surfaces. This enhancement could be significant for real-time usage before we get AVX or SSE in. For instance, in a test case with an 800x800 resolution, the frame rate increased from 51 fps to 74 fps.

The idea is bypassing SDL’s functions for 4bpp surfaces, which are notably slow without any apparent reason. Instead, I’ve replaced the get/set sections with faster versions, ensuring no loss of functionality.
![image](https://github.com/pygame-community/pygame-ce/assets/103119829/07e4e1a3-d357-4f2b-b652-24314879af17)

Starbuck5 commented:
Itzpr brought up a bug on discord where only changing lightness would lead to unexpected results. I see he's given this PR a "dont merge" label to reflect this, but I'm going to move it to draft just to take it off the review docket while this is resolved.

itzpr3d4t0r commented:
I have an update to share on this. After some experiments I found out that the current strategy for lightness isn't actually bugged but rather it's considered a naive way of handling lightness change. Take this picture for example:
![base2](https://github.com/pygame-community/pygame-ce/assets/103119829/12e4b124-947d-4482-8eb1-9f6c6734375d)
If we use my algorithm for just increasing its lightness by 60% we get this:
![base_pygame](https://github.com/pygame-community/pygame-ce/assets/103119829/46ece4ea-d98c-402c-a4f4-b1f68f00677b)
Wich seems weird but it's actually correct.

If we take an image manipulation software such as paint.net (or Krita) and we increase the surface lightness by 60% we get this:
![base_paint](https://github.com/pygame-community/pygame-ce/assets/103119829/26a57de2-0ba1-42ae-a2c1-331c89f3119b)

Now, fortunately I've found out that Krita offers 2 modes for changing lightness and one is so called "legacy". The lòegagy one actually matches up with my implementation's result:
![image](https://github.com/pygame-community/pygame-ce/assets/103119829/6f823c4a-f44b-49da-a9ab-f5a2dce0e4d9)

Fortunately Krita is open source so I can look up the implementation and correct this issue now.

Gabryel-lima reviewed (approved):
Wow, very good!
